Add config datalad.ui.interactive and allow non-interactive special remotes

bpoldrack opened the pull request:
- Makes `is_interactive` configurable via `datalad.ui.interactive`. Defaults to current behavior of detection via `isatty()`. The detection result is propagated to datalad subprocesses via env var (ping #7352).
- Adds a non-dialog annex backend variant and makes the UISwitcher use it, based on `is_interactive` when asked to set the `annex` backend.

Closes #7345
Closes #7349

mih reviewed (commented):
Just flew over the code, so this may be a bit too rough to be useful

bpoldrack commented on datalad/__init__.py:
```diff
@@ -118,12 +121,19 @@ def enable_librarymode():
+is_interactive = ut_is_interactive()
+# Propagate to subprocesses:
+os.environ['DATALAD_UI_INTERACTIVE'] = 'true' if is_interactive else 'false'
+
```
> but datalad code in a subprocess also runs this very code itself.

Yes, but suggestion is not entirely true. A python API user can set a config override. This would be lost on a subprocess. Very much like `enable_librarymode`.
Of course, such caller can set an env var themselves, but it requires awareness of possible datalad subprocesses (not only special remotes, but `sshrun`, git-credential-help, ..). I think, that's a needless burden.

bpoldrack replied:
Actually, just realized, I didn't actually solve my aim here. Will not resolve yet.

bpoldrack replied:
I made #7352 to address the general case instead.

bpoldrack replied:
Removed propagation from here in favor of above mentioned issue.

bpoldrack commented on datalad/interface/common_cfg.py:
```diff
@@ -640,6 +643,14 @@ def get_default_ssh():
+            'title': 'Datalad is executed interactively',
+            'text': "If enabled, datalad may prompt for user input. Disable for"
+                    "non-interactive usage."}),
+        'default': is_interactive_failsafe(),
```
Agree.

yarikoptic commented:
with such extensive change touching functionality I also wonder if may be worth/safer for master?

bpoldrack commented:
> with such extensive change touching functionality I also wonder if may be worth/safer for master?

Main reason for `maint` to me is, that we need this released ASAP.

bpoldrack commented:
I think, this is ready, @yarikoptic. There's the `git-annex died of signal 11` on macos github action, but this seems unrelated.

yarikoptic commented on datalad/utils.py:
```diff
@@ -369,14 +369,25 @@ def _is_stream_tty(stream: Optional[IO]) -> bool:
-    """
-    return all(_is_stream_tty(s) for s in (sys.stdin, sys.stdout, sys.stderr))
+
+def is_interactive_failsafe() -> bool:
```
please add docstring to describe why we need this failsafe?

also again -- code/logic duplication.  Please e.g. rename original implementation into `_is_interactive` or `is_interactive_noconfig` or add `check_config=True` and overload for the config default?

the point is to not duplicate logic which might need fixing in multiple spots later.  also note that we have `@never_fail`  so may be of use directly within invoking in config default?

bpoldrack replied:
Was unaware of never_fail indeed. Thx.

bpoldrack replied:
Not sure, what duplication you are referring to, though.

bpoldrack replied:
Just double-checked, but there's no duplication, I can see. `is_interactive()` now returns the config and the config gets its default from the new function, which is based on the former logic of `is_interactive`. It's a move of code, not duplication. What's the value of having that single line factored out again?

yarikoptic replied:
I am sorry -- I think I have misread the diff, indeed no duplication, sorry about that noise.

bpoldrack replied:
np

yarikoptic reviewed (changes requested):
- would break current interactions with users while installing/getting data from some datasets which do require authentication (e.g. i tried `datalad install -g ///crcns/aa-1`).  I feel that there should be some "deeper" fix for interactivity detection or `annex-no-dialog` should be used only when explicitly requested with config setting (e.g. if explicitly asked to be non-interactive)
  - note that we have `datalad.tests.ui.backend` config to control which UI backend to use during tests but we seems to have no `datalad.ui.backend`
- for a bug fix, PR IMHO unnecessarily changes API by flipping from `is_interactive` function to a module attribute, and now confusingly imported from various levels (`datalad` and `datalad.ui`).  Was there really a need for such a change if we are aiming for fixing some issue? (made review harder)
- code duplication should be reduced IMHO

bpoldrack commented on datalad/ui/__init__.py:
```diff
@@ -81,7 +83,9 @@ def set_backend(self, backend):
-                backend = 'dialog' if is_interactive() else 'no-progress'
+                backend = 'dialog' if is_interactive else 'no-progress'
+        if not is_interactive and backend == 'annex':
+            backend = 'annex-no-dialog'
```
> unfortunately this would break entirely ability of users to authenticate when possible

It doesn't break the ability, it changes it - you'd need to set the new config (as env var ATM). Because we are unable to correctly detect. And as far as I can tell, it's simply not possible. (see #7345)
Keeping the behavior of assuming interactivity by default, is not an option IMO, though. Because in non-interactive jobs, that just means stalling with no error reporting whatsoever. Users would have no clue at all.
Whereas in your case, there should at least be a hint that authentication wasn't possible. I'd rather err on the somewhat informative side than on the one that leaves users with nothing to act upon.

We could maintain the behavior to some extend, when we address #7352. Passing down the (accidently correct) detection from the super process, would lead to the same behavior in your case. I'd still argue it's wrong, since detection outside of a special remote process is not more correct. It's just slightly rarer to be an issue. (see #7354)

yarikoptic replied:
> It doesn't break the ability enitrely, it changes it - you'd need to set the new config (as env var ATM).

well, *not entirely* means that it still breaks it regardless how you name it.

Not breaking *at all* would be whenever the default behavior stays as is at least for `maint`. Eg. could be done via config variable staying "auto" by default and then switch to `True` or `False` based on the `utils.is_interactive()`, while keeping using `is_interactive()` function, and allowing all CIs or whatnot which knows that there is no agent behind keyboard to say so.  That would IMHO be better since it would be explicit and useful regardless on how we decide to trick/handle interactivity.    Then in `master` we could be deliberated further or even change behavior.

> Whereas in your case, there should at least be a hint that authentication wasn't possible.

why *wasn't possible* if it was and was working just fine before this change?

bpoldrack replied:
> Eg. could be done via config variable staying "auto" by default and then switch to True or False based on the utils.is_interactive(), while keeping using is_interactive() function

That is what is happening. The special remote process evaluates the detection in the state of this PR. But detection (`auto`) simply doesn't work. It only happens to do the right thing, when the process in question is the top-level process. But we are querying from within the special remote process.

> why wasn't possible if it was and was working just fine before this change?

I didn't "work" before that change. It happened to do the right thing in interactive sessions and the wrong thing in non-interactive ones, b/c it always assumed to be in an interactive one.

bpoldrack commented on datalad/__init__.py:
```diff
@@ -14,6 +14,9 @@
 From an interactive Python session, import `datalad.api` and inspect its
 documentation with `help`.
 """
+from __future__ import annotations
```
That's a left-over. Will remove.

bpoldrack commented:
@yarikoptic 
> I feel that there should be some "deeper" fix for interactivity detection 

I spend hours on this and couldn't find one. That's why this solution.

> or annex-no-dialog should be used only when explicitly requested with config setting

As I pointed out in the other spot: That means we are defaulting to stalling of non-interactive jobs. 

Also: Does that mean, you want "regular" datalad processes to default to (broken) detection, but special remote process to default to assuming interactivity? This would mean to completely ignore, that this issue is not something special about special remotes. Any datalad process can find itself in the exact same situation as a special remote process.

The mitigation I can see, is #7352 - passing config manager's state down to subprocesses. So, if the top-level datalad process is set to default (detection) the result of that detection would be set for the special remote process as well.

bpoldrack commented on datalad/ui/dialog.py:
```diff
@@ -357,6 +357,32 @@ def get_progressbar(self, *args, **kwargs):
+            kwargs['remote'] = self.specialremote
+            return AnnexSpecialRemoteProgressBar(*args, **kwargs)
+        else:
+            return super(NoDialogUnderAnnexUI, self).get_progressbar(
```
Ok

yarikoptic commented on datalad/ui/dialog.py:
```diff
@@ -332,29 +332,50 @@ def get_progressbar(self, *args, **kwargs):
+            return super().get_progressbar(*args, **kwargs)
+
+
+# poor man thingie for now
```
in what sense it is `poor`? looks legit to me

bpoldrack replied:
Not cleaned up, yet. That's your comment on the (moved) implementation of `UnderAnnexUI` ;-)

yarikoptic replied:
ah, LOL ;) so it is one of the skeletons ! ;)

yarikoptic commented on datalad/ui/tests/test_dialog.py:
```diff
@@ -173,25 +173,26 @@ def test_silent_question():
-
-@patch("datalad.log.is_interactive", lambda: False)
+    from ..dialog import SilentConsoleLog, NoDialogUnderAnnexUI
+    for k in [SilentConsoleLog, NoDialogUnderAnnexUI]:
```
best to move that outside of the test as `@pytest.mark.parametrize`:
- minimizes `diff`
- fail in one doesn't taint the run of the other
- allows to run those two runs in parallel or selecting one particular one

yarikoptic commented:
> The mitigation I can see, is #7352 - passing config manager's state down to subprocesses. So, if the top-level datalad process is set to default (detection) the result of that detection would be set for the special remote process as well (instead of running it's own detection which will always come back `False`).

yes -- if outside process `datalad.ui.interactive` set to `"auto"` , and it determines that it is interactive (or not), setting its env var `DATALAD_UI_INTERACTIVE` so any nested process also realizes that sounds like a viable way to go forward which would satisfy both of our desires?

> As I pointed out in the other spot: That means we are defaulting to stalling of non-interactive jobs.

... until realizing that the first stalled job is needing credentials (which it prompts for) and making sure that they have them in the future run which anyways would seems to be needed to do.

Again -- the other end of spectrum this PR introduces is that there is not prompt  to the user who (may be due to incorrect assumptions and defaults) was getting a prompt and now all of a sudden would just get a crash that it cannot get to that data, and no instructions even on how to mitigate, or to mitigate by explicitly saying "be interactive" in the interactive shell.

bpoldrack committed:
Add datalad.ui.interactive config and non-interactive backend for
special remotes

This patch introduces the config `datalad.ui.interactive` in order to
let users decide whether or not to run in interactive mode.
The config defaults to the former detection, except that this detection
is additionally safeguarded - any exception during that detection will
lead to non-interactive mode. In addition, the result of the detection
will be stored in `DATALAD_UI_INTERACTIVE`, thus passing down the result
to possible subprocesses rather than having them running their own
detection. Ultimately, this is about `ui`'s ability to talk to the
terminal via `getpass` and the detection does not work from within
subprocesses.

Closes #7349

Furthermore, this adds a non-interactive UI backend for annex special
remotes, which previously always assumed to be in interactive mode, and
adds the respective capacity for the UI_Switcher.

Closes #7345

bpoldrack commented:
@yarikoptic 
> yes -- if outside process datalad.ui.interactive set to "auto" , and it determines that it is interactive (or not), setting its env var DATALAD_UI_INTERACTIVE so any nested process also realizes that sounds like a viable way to go forward which would satisfy both of our desires?

Here is a test commit: https://github.com/datalad/datalad/pull/7344/commits/5417429ed9f004091e1ec794aa793b4793f6d820

> would need to set DATALAD_UI_INTERACTIVE envvar in any case, so config based options are overriden too in the children processes uniformly.

Not sure, whether I parse that right. You mean, you'd need the env var to overrule anything set in config files? Yes, that, or `-c` or `datalad.cfg.set(..., scope='override')` (With above commit would be passed down either way)

yarikoptic commented:
> Not sure, whether I parse that right. You mean, you'd need the env var to overrule anything set in config files? Yes, that, or `-c` or `datalad.cfg.set(..., scope='override')` (With above commit would be passed down either way)

in effect - yes, it would overrule in the children processes, but not in the parent process where it would actually obtain initial value via any available means (config, env var, overwrites) and then set explicit bool value it into `DATALAD_UI_INTERACTIVE` so all subprocesses (and this parent process later on may be) use it. So,  I do not immediately see any problem with the fact that children processes will not use value from the config value, and do not see a reason to utilize `DATALAD_CONFIG_OVERRIDES_JSON` for this specific aspect.

Note: this config is asked from the general `datalad.cfg`, i.e. not per dataset `config`.

bpoldrack commented:
@yarikoptic 

> Note: this config is asked from the general datalad.cfg, i.e. not per dataset config.

Yes, I added this aspect to the changelog. `Dataset.repo` evaluation from within the runner seems to much of a performance hit for this feature. I guess, the need to have something passed to a subprocess, while requiring it to be specific to a dataset, is rare - can't think of anything ATM. So, currently I think this is the best  of all the imperfect solutions.

bpoldrack commented:
Verdict from devcall: Rip out generic passing down of configs again. Should only apply to `datalad.ui.interactive` for now.

bpoldrack commented:
FTR: Remaining failures in crippledFS and macos github actions look as if https://github.com/datalad/datalad/pull/7367 wasn't merged.

Could it be that restarting github actions is running on an outdated, cached merge commit while Travis and AppVeyor build anew?

mih commented:
Before this is merged (in particular to runner adjustment), I recommend looking at https://github.com/datalad/datalad-next/issues/325 for an alternative -- which seems simpler, and also more effective.

yarikoptic reviewed (commented):
some comments I found pending

yarikoptic commented on datalad/utils.py:
```diff
@@ -369,14 +369,36 @@ def _is_stream_tty(stream: Optional[IO]) -> bool:
+    try:
+        interactive = all(_is_stream_tty(s)
+                          for s in (sys.stdin, sys.stdout, sys.stderr))
+    except Exception as e:
```
I would worry about overall capturing of all exceptions here -- e.g. if detection breaks with new python or something like that -- we would miss it. Needs investigation of git history if more information is provided or just go back to no exception handling for now IMHO.

yarikoptic commented on datalad/utils.py:
```diff
@@ -369,14 +369,36 @@ def _is_stream_tty(stream: Optional[IO]) -> bool:
+        # Raise log level to DEBUG in this case, though.
+        CapturedException(e, level=logging.DEBUG)
+        interactive = False
+    os.environ['DATALAD_UI_INTERACTIVE'] = str(interactive)
```
I do not think it should overload if it is already defined.

yarikoptic commented on datalad/utils.py:
```diff
@@ -369,14 +369,36 @@ def _is_stream_tty(stream: Optional[IO]) -> bool:
 def is_interactive() -> bool:
-    """Return True if all in/outs are open and tty.
+    from datalad import cfg
+    return cfg.obtain("datalad.ui.interactive")
```
what if config is within local git repo config?

shouldn't we overload that DATALAD_UI_INTERACTIVE here? and if we are to decide for "auto" to be default -- then here call is_interactive_failsafe explicitly?

yarikoptic reviewed (commented):
more comments
